### PR TITLE
[YAML] Fix simple YAML mappings type hinting

### PR DESCRIPTION
**Pull request by Polber**
There have been type inferencing issues with Beam YAML `MapToFields` where the type(s) from the upstream pcollection are dropped when the `beam.Select` is used on unmapped fields in the `MapToFields` expansion (i.e. when using `append: true`, the unmapped fields are mapped to AnyType). 

The fix is to register type hints for the implicitly appended fields.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Polber** [comment]
R: @robertwb

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 1]
```diff
@@ -276,12 +277,19 @@ def _as_callable_for_pcoll(
...
 
 
-def _as_callable(original_fields, expr, transform_name, language):
+def _as_callable(
+    original_fields, expr, transform_name, language, input_schema=None):
+
+  # Extract original type from upstream pcoll when doing simple mappings
+  if input_schema is None:
+    input_schema = {}
+  original_type = input_schema.get(transform_name, None)
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_mapping.py:290]: I'm not quite understanding why you want to use `transform_name` as a key in the input schema here.

Comment (Polber) [reply sdks/python/apache_beam/yaml/yaml_mapping.py]: There is no good reason, I think that was the result of an earlier null issue that no longer applies. Removed a couple lines related to obsolete cases

**Review by robertwb — APPROVED** [review]
It'd be good to add a test if it's not that hard.

**Comment by Polber** [comment]
> It'd be good to add a test if it's not that hard.

@robertwb 
Added a test.

Interestingly, if I remove the following from the test, it correctly inferences the types...
```
elements.element_type = schemas.named_tuple_from_schema(
    schema_pb2.Schema(
        fields=[
            schemas.schema_field('label', str),
            schemas.schema_field('conductor', int),
            schemas.schema_field('rank', int)
        ]))
```
i.e. if exactly what is done in the ReadFromPubSub
https://github.com/apache/beam/blob/df067dba5dc509135e857596224759e485d47a0b/sdks/python/apache_beam/yaml/yaml_io.py#L325-L326

**Comment by Polber** [comment]
Fixes #31434

**Pull request merged**